File: go/canlog/canlog.go
package canlog

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type CanLog struct {
	logs     []*log
	formater FormatFunc
}

// Canlog型はCanlogに合わせたデータ構造を持ったオブジェクト
type log struct {
	time   float64
	ch     uint8
	id     uint16
	dir    string
	stat   string
	dlc    uint8
	data   []byte
	remain string
}

type FormatFunc func(time float64, ch uint8, id uint16, dir string, dlc uint8, data []byte) string

// NewCanLog は、新たにCanlogオブジェクトを作成する
func NewCanLog() *CanLog {
	return &CanLog{
		formater: formatLog,
	}
}

// NewLog は、新たにlogオブジェクトを作成する
func newLog() *log {
	return &log{}
}

// Parseは、ファイルを解析してCanlogオブジェクトを作成する
func (cl *CanLog) Parse(filename string) error {
	fp, err := os.Open(filename)
	if err != nil {
		return err
	}
	defer fp.Close()

	scanner := bufio.NewScanner(fp)

	for scanner.Scan() {
		l := newLog()

		fs := strings.Fields(scanner.Text())
		if fs[1] == "1" || fs[1] == "2" {
			l.time, err = strconv.ParseFloat(fs[0], 32)
			if err != nil {
				return err
			}
			c, err := strconv.ParseUint(fs[1], 10, 32)
			if err != nil {
				return err
			}
			l.ch = uint8(c)
			i, err := strconv.ParseUint(fs[2], 16, 32)
			if err != nil {
				return err
			}
			l.id = uint16(i)
			l.dir = fs[3]
			l.stat = fs[4]
			d, err := strconv.Atoi(fs[5])
			if err != nil {
				return err
			}
			l.dlc = uint8(d)
			for _, f := range fs[6 : l.dlc+6] {
				d, _ := strconv.ParseUint(f, 16, 32)
				l.data = append(l.data, byte(d))
			}
			l.remain = strings.Join(fs[l.dlc+7:l.dlc+15], " ")
			cl.logs = append(cl.logs, l)
		}
	}
	return scanner.Err()
}

// PickRecordは、引数で渡したCanlogオブジェクトからidsで渡したIDが存在するレコードのみを抽出したCanlogオブジェクトを返す
func PickRecord(c *CanLog, ids []uint16) *CanLog {
	nc := NewCanLog()
	for _, r := range c.logs {
		if isContains(r.id, ids) {
			nc.logs = append(nc.logs, r)
		}
	}
	return nc
}

// DelRecordは、引数で渡したCanlogオブジェクトからidsで渡したIDが存在するレコードを削除したCanlogオブジェクトを返す
func DelRecord(c *CanLog, ids []uint16) *CanLog {
	nc := NewCanLog()

	for _, r := range c.logs {
		if !isContains(r.id, ids) {
			nc.logs = append(nc.logs, r)
		}
	}
	return nc
}

// isContainsは、tgtが、配列arrに含まれているかをbooleanとして返す
func isContains(tgt uint16, arr []uint16) bool {
	for _, e := range arr {
		if e == tgt {
			return true
		}
	}
	return false
}

// PrintLogは、CanLogオブジェクトを1レコードずつフォーマットして標準出力へ出力する
func (c *CanLog) String() string {
	s := ""
	for _, l := range c.logs {
		s += c.formater(l.time, l.ch, l.id, l.dir, l.dlc, l.data)
	}
	return s
}

func formatLog(time float64, ch uint8, id uint16, dir string, dlc uint8, data []byte) string {
	s := bytesToSeparatedString(data, " ")
	return fmt.Sprintf("%06f %s %03s %s %X %s\n", time, ch, id, dir, dlc, s)
}

// bytesToSeparatedStringは、byte sliceをsepで区切った文字列へ変換する
func bytesToSeparatedString(bytes []byte, sep string) string {
	var data []string

	for _, b := range bytes {
		data = append(data, fmt.Sprintf("%02X", b))
	}
	return strings.Join(data, sep)
}

func (c *CanLog) Format(formater FormatFunc) {
	c.formater = formater
}

File: perl/canlog/canlog.pl
package main;

use 5.008001;
use strict;
use warnings;
use utf8;

use Getopt::Kingpin;

my $kingpin = Getopt::Kingpin->new;
my $filename = $kingpin->arg("file", "input log file")->string;
$kingpin->parse;

my $canlog = CanLog->new;
$canlog->parse($filename);
print $canlog->string . "\n";

$canlog->stringer(sub{
    my ($obj, $sp) = @_;
    my $str .= sprintf "%f%s%03X", $obj->{TIME} / 1000000, $sp, $obj->{ID};
    $str .= sprintf "%s%02X", $sp, $_ foreach @{$obj->{DATA}};
    return $str
});
print $canlog->string . "\n";

my $timediff = $canlog->timediff;
my $strings  = $canlog->strings;
for (my $i = 0; $i< scalar(@$strings); $i++) {
    printf "%f %s\n", @$timediff[$i] / 1000000, @$strings[$i];
}

package CanLog;

use 5.008001;
use strict;
use warnings;
use Path::Tiny;

sub new {
    my $class = shift;
    my $self = {};
    $self->{LOGS} = ();
    $self->{SPLITER}  = " ";
    $self->{STRINGER} = \&string_simple;
    return bless $self, $class;
}

sub parse {
    my ($self, $filename) = @_;
    my $path = path($filename);
    foreach my $line ($path->lines) {
        chomp($line);
        if ($line =~ /^\s+\d+\.\d+\s[1-9]/) {
            my $log = Log->new;
            $log->parse($line);
            push @{$self->{LOGS}}, $log;
        }
    }
}

sub timediff {
    my $self = shift;
    my @diff = ();
    my $prev = 0000000;
    foreach my $log (values($self->{LOGS})) {
        push @diff, $log->{TIME} - $prev;
        $prev = $log->{TIME};
    }
    return \@diff;
}

sub strings {
    my $self = shift;
    my @strings = ();
    foreach my $log (values($self->{LOGS})) {
        push @strings, $log->string($self->{STRINGER}, $self->{SPLITER});
    }
    return \@strings;
}

sub string {
    my $self = shift;
    my $string = "";
    foreach my $log (values($self->{LOGS})) {
        $string .= $log->string($self->{STRINGER}, $self->{SPLITER}) . "\n";
    }
    return $string;
}

sub string_simple {
    my ($obj, $sp) = @_;
    my $str .= sprintf "%f%s%d%s%03X%s%s%s%X", $obj->{TIME} / 1000000, $sp, $obj->{CH},
        $sp, $obj->{ID}, $sp, $obj->{DIR} eq Log->RX ? "Rx" : "Tx", $sp, $obj->{DLC};
    $str .= sprintf "%s%02X", $sp, $_ foreach @{$obj->{DATA}};
    return $str
}

sub spliter {
    my ($self, $sp) = @_;
    $self->{SPLITER} = $sp;
}

sub stringer {
    my ($self, $subroutine) = @_;
    $self->{STRINGER} = $subroutine;
}

package Log;

use 5.008001;
use strict;
use warnings;

use constant {
    RX => 0,
    TX => 1,
};

sub new {
    my $class = shift;
    my $self = {};
    $self->{TIME}     = undef;
    $self->{CH}       = undef;
    $self->{ID}       = undef;
    $self->{DIR}      = undef;
    $self->{STAT}     = undef;
    $self->{DLC}      = undef;
    $self->{DATA}     = [];
    $self->{REMAIN}   = [];
    return bless $self, $class;
}

sub parse {
    my ($self, $line) = @_;
    my @fields = split(/\s+/, $line);
    my ($time, $ch, $id, $dir, $stat, $dlc) = @fields[1..7];
    my @data = @fields[7..6+$dlc];
    my @remain = @fields[6+$dlc..scalar(@fields)-1];

    $self->{TIME}    = int($time * 1000000);
    $self->{CH}      = int($ch);
    $self->{ID}      = hex($id);
    $self->{DIR}     = $dir eq "Rx" ? RX : TX;
    $self->{STAT}    = $stat;
    $self->{DLC}     = $dlc;
    @{$self->{DATA}} =  map(hex, @data);
    $self->{REMAIN}  = @remain;
}

sub string {
    my ($self, $stringer, $spliter) = @_;
    return &{$stringer}($self, $spliter);
}
